// bsv-sdk/es/utils/CommonUtils.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _regeneratorRuntime from "@babel/runtime/regenerator";
export function sleep(_x) {
  return _sleep.apply(this, arguments);
}

function _sleep() {
  _sleep = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(time) {
    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", new Promise(function (resolve, reject) {
              setTimeout(function () {
                resolve('');
              }, time * 1000);
            }));

          case 1:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _sleep.apply(this, arguments);
}

export function reverceFtByteString(str) {
  str = str.substr(0, str.length - 8);
  var ret = '';

  for (var i = 0; i < str.length; i += 2) {
    ret = str[i] + str[i + 1] + ret;
  }

  return ret;
}
export function getMergeTime(count, mergeTime) {
  if (mergeTime === void 0) {
    mergeTime = 0;
  }

  var utxoCount = count;

  if (utxoCount <= 20) {
    return [utxoCount, mergeTime];
  } else {
    utxoCount -= 19;
    mergeTime++;
    return getMergeTime(utxoCount, mergeTime);
  }
}
export var FT = {
  "MC": {
    codehash: "777e4dd291059c9f7a0fd563f7204576dcceb791",
    genesis: "54256eb1b9c815a37c4af1b82791ec6bdf5b3fa3",
    genesisTxid: "fdc76b124e88f345a4400d7408588bf5dca8d3e8c8c4c306d6d4a77f1df8043e",
    sensibleId: "3e04f81d7fa7d4d606c3c4c8e8d3a8dcf58b5808740d40a445f3884e126bc7fd00000000"
  },
  "SHOW": {
    codehash: "777e4dd291059c9f7a0fd563f7204576dcceb791",
    genesis: "54754f357afafa9e77488de7135f403276e03a9b",
    genesisTxid: "6d5804bc1a433df8d1200fd28541e736a96452de311a87d088b30aa8cb8d3712",
    sensibleId: "12378dcba80ab388d0871a31de5264a936e74185d20f20d1f83d431abc04586d00000000"
  }
};
