// sensible-sdk/dist/common/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrInfo = exports.ErrCode = exports.CodeError = void 0;
class CodeError extends Error {
    constructor(code, msg) {
        super(msg);
        this.code = code;
        if (msg) {
            this.message = msg;
        }
        else {
            this.message = "CodeError:" + code;
        }
    }
}
exports.CodeError = CodeError;
var ErrCode;
(function (ErrCode) {
    ErrCode[ErrCode["EC_OK"] = 0] = "EC_OK";
    ErrCode[ErrCode["EC_INNER_ERROR"] = -1] = "EC_INNER_ERROR";
    ErrCode[ErrCode["EC_INVALID_ARGUMENT"] = -2] = "EC_INVALID_ARGUMENT";
    ErrCode[ErrCode["EC_SENSIBLE_API_ERROR"] = -3] = "EC_SENSIBLE_API_ERROR";
    ErrCode[ErrCode["EC_INVALID_SIGNERS"] = -4] = "EC_INVALID_SIGNERS";
    //需要特殊处理
    ErrCode[ErrCode["EC_UTXOS_MORE_THAN_3"] = -100] = "EC_UTXOS_MORE_THAN_3";
    ErrCode[ErrCode["EC_TOO_MANY_FT_UTXOS"] = -101] = "EC_TOO_MANY_FT_UTXOS";
    ErrCode[ErrCode["EC_FIXED_TOKEN_SUPPLY"] = -102] = "EC_FIXED_TOKEN_SUPPLY";
    //金额不足
    ErrCode[ErrCode["EC_INSUFFICIENT_BSV"] = -200] = "EC_INSUFFICIENT_BSV";
    ErrCode[ErrCode["EC_INSUFFICIENT_FT"] = -201] = "EC_INSUFFICIENT_FT";
    ErrCode[ErrCode["EC_NFT_NOT_ON_SELL"] = -300] = "EC_NFT_NOT_ON_SELL";
})(ErrCode = exports.ErrCode || (exports.ErrCode = {}));
exports.ErrInfo = {
    InvalidArgument: {
        code: ErrCode.EC_INVALID_ARGUMENT,
        message: "",
    },
};
